### ISSUE TITLE
How to handle conditionals in list items

### ISSUE THREAD

**Issue by spockNinja** [awslabs/cfn-python-lint#350]
*cfn-lint version: `0.7.1`*

*I have a condition around whether or not a fargate service uses the service registry. The way I had it set up fails to pass linting, but does work in CloudFormation.*

Here is the template causing the problem (abbreviated to just the problem areas):

```yaml
AWSTemplateFormatVersion: 2010-09-09

Parameters:

  ServiceDiscoveryName:
    Description: Name of the service when used for internal service discovery
    Type: String
    Default: ''

Conditions:

  UseServiceDiscovery:
    Fn::Not:
      - Fn::Equals:
          - Ref: ServiceDiscoveryName
          - ''

Resources:

  Service:
    Type: AWS::ECS::Service
    Properties:
      # other properties here
      ServiceRegistries:
        Fn::If:
          - UseServiceDiscovery
          - - RegistryArn:
                Fn::GetAtt: [ServiceDiscoveryRegistry, Arn]
          - Ref: AWS::NoValue
```

This results in the linter error:
```
Property ServiceRegistries should be of type List for resource Service at Resources/Service/Properties/ServiceRegistries
cloudformation/shared-templates/fargate-service-alb.yaml:111:7
```

I then tried the following change:
```
    ServiceRegistries:
        - RegistryArn:
            Fn::If:
              - UseServiceDiscovery
              - Fn::GetAtt: [ServiceDiscoveryRegistry, Arn]
              - Ref: AWS::NoValue
```

And that did appease the linter. However, it resulted in the following CloudFormation error:
```
registry arn cannot be blank. (Service: AmazonECS; Status Code: 400; Error Code: InvalidParameterException; Request ID: 5bd193f7-b861-11e8-8b0a-cd1b086b8bb9)
```

Our work-around to keep us moving was to add `Ref` to our spec override:

```json
    "AWS::ECS::Service.ServiceRegistry": {
      "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-serviceregistry.html",
      "Properties": {
        "Port": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-serviceregistry.html#cfn-ecs-service-serviceregistry-port",
          "PrimitiveType": "Integer",
          "Required": false,
          "UpdateType": "Mutable"
        },
        "RegistryArn": {
          "Documentation": "http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-serviceregistry.html#cfn-ecs-service-serviceregistry-registryarn",
          "PrimitiveType": "String",
          "Required": false,
          "UpdateType": "Mutable"
        },
        "Ref": {
          "PrimitiveType": "String",
          "Required": false
        }
      }
```

and then meet in the middle with:

```
      ServiceRegistries:
        Fn::If:
          - UseServiceDiscovery
          - - RegistryArn:
                Fn::GetAtt: [ServiceDiscoveryRegistry, Arn]
          - - Ref: AWS::NoValue
```

I do believe that final template makes the most sense to appease the linter and CloudFormation. But without the spec override, we get:
```
E3002 Invalid Property Resources/Service/Properties/ServiceRegistries/0/Ref
cloudformation/shared-templates/fargate-service-alb.yaml:111:7
```

If I am doing something silly or missing something else, corrections are certainly welcome. If not, is there an easy way that `Ref: AWS::NoValue` could be respected as special and not cause an error when it is sitting in for another `Type`?

Thanks for this tool! I think it's going to greatly speed up our CloudFormation generation and testing. I really want to see it succeed. I am happy to provide any other context or input.

**Comment by kddejong** [comment]
This is a bug.  Working on a fix for it now and should have something shortly.  We aren't properly handling the Ref AWS::NoValue which is messing this up.  This was an earlier rule that could use a little updating.  While we aren't where I want to be with conditions we are a lot closer and should have been catching this.  Thank you!

**Comment by spockNinja** [comment]
That was quick! Thanks for the work here. I thoroughly appreciate how well this tool is being attended. 😄 💯
